File: MatrixCrypto.py
from Crypto import Crypto


class MatrixCripto(Crypto):
    @staticmethod
    def encrypt(message, key):
        mat = []
        for i in range(0, len(message), key):
            mat.append(list(message[i:i + key]))
        result = ''
        for i in range(key):
            for j in range(len(mat)):
                try:
                    result += mat[j][i]
                except IndexError:
                    result += ' '
        return result

    @staticmethod
    def decrypt(message, key):
        mat = []
        col = len(message) // key
        for i in range(0, len(message), col):
            mat.append(list(message[i:i + col]))
        result = ''
        for i in range(col):
            for j in range(len(mat)):
                try:
                    result += mat[j][i]
                except IndexError:
                    result += ' '
        return result.strip()


testData = 'mmmdddjjjj'
print('Test data: ' + testData)
print()

print('Matrix Cripto:')
matEncrypted = MatrixCripto.encrypt(testData, 3)
print('Encrypted text: ', matEncrypted)
print('Decrypted text: ', MatrixCripto.decrypt(matEncrypted, 3))

File: c++/MatrixCrypto.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class MatrixCripto {
public:
    static string encrypt(const string& message, int key) {
        vector<vector<char>> mat;
        string result = "";

        for (size_t i = 0; i < message.length(); i += key) {
            vector<char> row(message.begin() + i, message.begin() + min(i + key, message.length()));
            mat.push_back(row);
        }

        for (int i = 0; i < key; ++i) {
            for (size_t j = 0; j < mat.size(); ++j) {
                if (i < mat[j].size()) {
                    result += mat[j][i];
                }
                else {
                    result += ' ';
                }
            }
        }

        return result;
    }

    static string decrypt(const string& message, int key) {
        vector<vector<char>> mat;
        string result = "";
        int col = message.length() / key;

        for (size_t i = 0; i < message.length(); i += col) {
            vector<char> row(message.begin() + i, message.begin() + min(i + col, message.length()));
            mat.push_back(row);
        }

        for (int i = 0; i < col; ++i) {
            for (size_t j = 0; j < mat.size(); ++j) {
                if (i < mat[j].size()) {
                    result += mat[j][i];
                }
                else {
                    result += ' ';
                }
            }
        }

        return result;
    }
};

int main() {
    string testData = "mmmdddjjjj";
    cout << "Test data: " << testData << endl;

    cout << "\nMatrix Cripto:" << endl;
    string matEncrypted = MatrixCripto::encrypt(testData, 3);
    cout << "Encrypted text: " << matEncrypted << endl;
    cout << "Decrypted text: " << MatrixCripto::decrypt(matEncrypted, 3) << endl;

    return 0;
}
